File: zig_rewrite/parser.zig
const std = @import("std");
const lexer = @import("lexer.zig");

const ParseError = error{
    UnexpectedToken,
    ExpectedIdent,
    ExpectedReg,
    ExpectedFunc,
    ExpectedType,
    ExpectedFuncName,
    UnexpectedExprToken,
    UnexpectedStmtToken,
    UnexpectedTopLevel,
    TooManyRegisters,
    OutOfMemory,
};

/// Per-function register name -> index map built during parsing.
const RegMap = struct {
    names: std.ArrayListUnmanaged([]const u8),
    alloc: std.mem.Allocator,

    fn init(alloc: std.mem.Allocator) RegMap {
        return .{ .names = .empty, .alloc = alloc };
    }

    fn deinit(self: *RegMap) void {
        self.names.deinit(self.alloc);
    }

    fn intern(self: *RegMap, name: []const u8) !lexer.RegIdx {
        for (self.names.items, 0..) |n, i|
            if (std.mem.eql(u8, n, name)) return @intCast(i);
        const idx: lexer.RegIdx = @intCast(self.names.items.len);
        if (idx == std.math.maxInt(lexer.RegIdx)) return error.TooManyRegisters;
        try self.names.append(self.alloc, name);
        return idx;
    }

    fn count(self: *const RegMap) u16 {
        return @intCast(self.names.items.len);
    }
};

const Parser = struct {
    tokens: []const lexer.Token,
    pos: usize,
    alloc: std.mem.Allocator,
    slab: *lexer.ExprSlab,

    fn peek(self: *Parser) lexer.Token {
        return self.tokens[self.pos];
    }

    fn advance(self: *Parser) lexer.Token {
        const t = self.tokens[self.pos];
        if (self.pos + 1 < self.tokens.len) self.pos += 1;
        return t;
    }

    fn expectTag(self: *Parser, tag: lexer.TokenTag) ParseError!lexer.Token {
        const t = self.advance();
        if (std.meta.activeTag(t) == tag) return t;
        return error.UnexpectedToken;
    }

    fn expectIdent(self: *Parser) ParseError![]const u8 {
        return switch (self.advance()) {
            .ident => |s| s,
            else => error.ExpectedIdent,
        };
    }

    fn expectReg(self: *Parser) ParseError![]const u8 {
        return switch (self.advance()) {
            .reg => |s| s,
            else => error.ExpectedReg,
        };
    }

    fn expectFunc(self: *Parser) ParseError![]const u8 {
        return switch (self.advance()) {
            .func => |s| s,
            else => error.ExpectedFunc,
        };
    }

    fn parseTy(self: *Parser) ParseError!lexer.Ty {
        return switch (self.advance()) {
            .ty_int => .int,
            .ty_void => .void_,
            .ty_string => .str,
            .ty_bool => .bool_,
            .ident => .named,
            else => error.ExpectedType,
        };
    }

    fn parseArgs(self: *Parser, rm: *RegMap) ParseError!std.ArrayList(*lexer.Expr) {
        var args: std.ArrayList(*lexer.Expr) = .empty;
        if (std.meta.activeTag(self.peek()) != .lparen) return args;
        _ = self.advance();
        while (std.meta.activeTag(self.peek()) != .rparen) {
            try args.append(self.alloc, try self.parseExpr(rm));
            if (std.meta.activeTag(self.peek()) == .comma) _ = self.advance();
        }
        _ = try self.expectTag(.rparen);
        return args;
    }

    fn box(self: *Parser, e: lexer.Expr) ParseError!*lexer.Expr {
        return self.slab.alloc_node(e);
    }

    fn parseBinOp(self: *Parser, rm: *RegMap, comptime tag: lexer.TokenTag) ParseError!*lexer.Expr {
        _ = self.advance();
        const a = try self.parseExpr(rm);
        _ = try self.expectTag(.comma);
        const b = try self.parseExpr(rm);
        const payload = lexer.BinOp{ .a = a, .b = b };
        return self.box(switch (tag) {
            .kw_add => .{ .add = payload },
            .kw_sub => .{ .sub = payload },
            .kw_mul => .{ .mul = payload },
            .kw_div => .{ .div = payload },
            .kw_lt => .{ .lt = payload },
            .kw_gt => .{ .gt = payload },
            .kw_eq => .{ .eq = payload },
            else => unreachable,
        });
    }

    fn parseExpr(self: *Parser, rm: *RegMap) ParseError!*lexer.Expr {
        return switch (self.peek()) {
            .kw_const => blk: {
                _ = self.advance();
                break :blk try self.parseExpr(rm);
            },
            .kw_add => try self.parseBinOp(rm, .kw_add),
            .kw_sub => try self.parseBinOp(rm, .kw_sub),
            .kw_mul => try self.parseBinOp(rm, .kw_mul),
            .kw_div => try self.parseBinOp(rm, .kw_div),
            .kw_lt => try self.parseBinOp(rm, .kw_lt),
            .kw_gt => try self.parseBinOp(rm, .kw_gt),
            .kw_eq => try self.parseBinOp(rm, .kw_eq),
            .kw_call => blk: {
                _ = self.advance();
                const name = switch (self.advance()) {
                    .ident => |s| s,
                    .func => |s| s,
                    else => return error.ExpectedFuncName,
                };
                break :blk try self.box(.{ .call = .{ .name = name, .args = try self.parseArgs(rm) } });
            },
            .kw_alloc => blk: {
                _ = self.advance();
                break :blk try self.box(.{ .alloc = try self.parseExpr(rm) });
            },
            .int => |n| blk: {
                _ = self.advance();
                break :blk try self.box(.{ .int = n });
            },
            .str => |s| blk: {
                _ = self.advance();
                break :blk try self.box(.{ .str = s });
            },
            .reg => |r| blk: {
                _ = self.advance();
                const idx = try rm.intern(r);
                if (std.meta.activeTag(self.peek()) == .dot) {
                    _ = self.advance();
                    break :blk try self.box(.{ .field = .{
                        .reg = idx,
                        .field = try self.expectIdent(),
                    } });
                }
                break :blk try self.box(.{ .reg = idx });
            },
            .ident => |name| blk: {
                _ = self.advance();
                if (std.meta.activeTag(self.peek()) == .lbrace) {
                    _ = self.advance();
                    var fields: std.ArrayList(lexer.FieldInit) = .empty;
                    while (std.meta.activeTag(self.peek()) != .rbrace) {
                        const fname = try self.expectIdent();
                        _ = try self.expectTag(.colon);
                        try fields.append(self.alloc, .{
                            .name = fname,
                            .expr = try self.parseExpr(rm),
                        });
                        if (std.meta.activeTag(self.peek()) == .comma) _ = self.advance();
                    }
                    _ = try self.expectTag(.rbrace);
                    break :blk try self.box(.{ .struct_lit = .{ .name = name, .fields = fields } });
                }
                break :blk try self.box(.{ .named = name });
            },
            else => error.UnexpectedExprToken,
        };
    }

    fn parseStmt(self: *Parser, rm: *RegMap) ParseError!lexer.Stmt {
        return switch (self.peek()) {
            .reg => |r| blk: {
                _ = self.advance();
                const idx = try rm.intern(r);
                _ = try self.expectTag(.assign);
                break :blk .{ .assign = .{ .reg = idx, .expr = try self.parseExpr(rm) } };
            },
            .kw_set => blk: {
                _ = self.advance();
                const r = try self.expectReg();
                const idx = try rm.intern(r);
                _ = try self.expectTag(.dot);
                const field = try self.expectIdent();
                _ = try self.expectTag(.assign);
                break :blk .{ .set_field = .{
                    .reg = idx,
                    .field = field,
                    .expr = try self.parseExpr(rm),
                } };
            },
            .kw_call => blk: {
                _ = self.advance();
                const name = switch (self.advance()) {
                    .ident => |s| s,
                    .func => |s| s,
                    else => return error.ExpectedFuncName,
                };
                break :blk .{ .call = .{ .name = name, .args = try self.parseArgs(rm) } };
            },
            .kw_ret => blk: {
                _ = self.advance();
                break :blk .{ .ret = try self.parseExpr(rm) };
            },
            .kw_while => blk: {
                _ = self.advance();
                _ = try self.expectTag(.lparen);
                const cond = try self.parseExpr(rm);
                _ = try self.expectTag(.rparen);
                _ = try self.expectTag(.lbrace);
                var body: std.ArrayList(lexer.Stmt) = .empty;
                while (std.meta.activeTag(self.peek()) != .rbrace)
                    try body.append(self.alloc, try self.parseStmt(rm));
                _ = try self.expectTag(.rbrace);
                break :blk .{ .while_ = .{ .cond = cond, .body = body } };
            },
            else => error.UnexpectedStmtToken,
        };
    }

    fn parseStruct(self: *Parser) ParseError!lexer.StructDef {
        _ = try self.expectTag(.kw_struct);
        const name = try self.expectIdent();
        _ = try self.expectTag(.lbrace);
        var fields: std.ArrayList(lexer.StructField) = .empty;
        while (std.meta.activeTag(self.peek()) != .rbrace) {
            const fname = try self.expectIdent();
            _ = try self.expectTag(.colon);
            try fields.append(self.alloc, .{ .name = fname, .ty = try self.parseTy() });
        }
        _ = try self.expectTag(.rbrace);
        return .{ .name = name, .fields = fields };
    }

    fn parseFunction(self: *Parser) ParseError!lexer.Function {
        const name = try self.expectFunc();
        var rm = RegMap.init(self.alloc);
        defer rm.deinit();

        var params: std.ArrayList(lexer.Param) = .empty;
        if (std.meta.activeTag(self.peek()) == .lparen) {
            _ = self.advance();
            while (std.meta.activeTag(self.peek()) != .rparen) {
                const pname = try self.expectReg();
                _ = try self.expectTag(.colon);
                const ty = try self.parseTy();
                const idx = try rm.intern(pname);
                try params.append(self.alloc, .{ .name = pname, .ty = ty, .idx = idx });
                if (std.meta.activeTag(self.peek()) == .comma) _ = self.advance();
            }
            _ = try self.expectTag(.rparen);
        }
        const ret_ty: lexer.Ty = if (std.meta.activeTag(self.peek()) == .colon) blk: {
            _ = self.advance();
            break :blk try self.parseTy();
        } else .void_;
        _ = try self.expectTag(.lbrace);
        var body: std.ArrayList(lexer.Stmt) = .empty;
        while (std.meta.activeTag(self.peek()) != .rbrace)
            try body.append(self.alloc, try self.parseStmt(&rm));
        _ = try self.expectTag(.rbrace);
        return .{
            .name = name,
            .params = params,
            .ret_ty = ret_ty,
            .body = body,
            .n_regs = rm.count(),
        };
    }

    fn parseProgram(self: *Parser) ParseError!lexer.Program {
        var structs: std.ArrayList(lexer.StructDef) = .empty;
        var functions: std.ArrayList(lexer.Function) = .empty;
        while (std.meta.activeTag(self.peek()) != .eof) {
            switch (self.peek()) {
                .kw_struct => try structs.append(self.alloc, try self.parseStruct()),
                .func => try functions.append(self.alloc, try self.parseFunction()),
                else => return error.UnexpectedTopLevel,
            }
        }
        return .{ .structs = structs, .functions = functions, .slab = undefined, .tokens = undefined };
    }
};

pub fn parse(tokens: []const lexer.Token, token_list: std.ArrayList(lexer.Token), alloc: std.mem.Allocator) !lexer.Program {
    var slab = try lexer.ExprSlab.init(alloc, tokens.len / 3 + 64);
    errdefer slab.deinit(alloc);
    var p = Parser{ .tokens = tokens, .pos = 0, .alloc = alloc, .slab = &slab };
    var prog = try p.parseProgram();
    prog.slab = slab;
    prog.tokens = token_list;
    return prog;
}

File: src/parser.rs
use crate::ast::*;
use crate::lexer::Token;

struct Parser {
    tokens: Vec<Token>,
    pos: usize,
}

impl Parser {
    fn peek(&self) -> &Token {
        &self.tokens[self.pos]
    }

    fn advance(&mut self) -> Token {
        let t = self.tokens[self.pos].clone();
        if self.pos + 1 < self.tokens.len() {
            self.pos += 1;
        }
        t
    }

    fn expect(&mut self, expected: &Token) -> Result<(), String> {
        let t = self.advance();
        if std::mem::discriminant(&t) == std::mem::discriminant(expected) {
            Ok(())
        } else {
            Err(format!("Expected {expected:?}, got {t:?}"))
        }
    }

    fn expect_ident(&mut self) -> Result<String, String> {
        match self.advance() {
            Token::Ident(s) => Ok(s),
            t => Err(format!("Expected identifier, got {t:?}")),
        }
    }

    fn expect_reg(&mut self) -> Result<String, String> {
        match self.advance() {
            Token::Reg(s) => Ok(s),
            t => Err(format!("Expected register (%name), got {t:?}")),
        }
    }

    fn expect_func(&mut self) -> Result<String, String> {
        match self.advance() {
            Token::Func(s) => Ok(s),
            t => Err(format!("Expected function (@name), got {t:?}")),
        }
    }

    fn parse_ty(&mut self) -> Result<Ty, String> {
        match self.advance() {
            Token::TyInt => Ok(Ty::Int),
            Token::TyVoid => Ok(Ty::Void),
            Token::TyString => Ok(Ty::Str),
            Token::TyBool => Ok(Ty::Bool),
            Token::Ident(s) => Ok(Ty::Named(s)),
            t => Err(format!("Expected type, got {t:?}")),
        }
    }

    /// Parse an optional argument list. Parens are optional — `call foo` is valid.
    fn parse_args(&mut self) -> Result<Vec<Expr>, String> {
        if self.peek() != &Token::LParen {
            return Ok(Vec::new());
        }
        self.advance();
        let mut args = Vec::new();
        while self.peek() != &Token::RParen {
            args.push(self.parse_expr()?);
            if self.peek() == &Token::Comma {
                self.advance();
            }
        }
        self.expect(&Token::RParen)?;
        Ok(args)
    }

    fn parse_expr(&mut self) -> Result<Expr, String> {
        match self.peek().clone() {
            Token::Const => {
                self.advance();
                let inner = self.parse_expr()?;
                Ok(Expr::Const(Box::new(inner)))
            }
            Token::Add => {
                self.advance();
                let a = self.parse_expr()?;
                self.expect(&Token::Comma)?;
                let b = self.parse_expr()?;
                Ok(Expr::Add(Box::new(a), Box::new(b)))
            }
            Token::Sub => {
                self.advance();
                let a = self.parse_expr()?;
                self.expect(&Token::Comma)?;
                let b = self.parse_expr()?;
                Ok(Expr::Sub(Box::new(a), Box::new(b)))
            }
            Token::Mul => {
                self.advance();
                let a = self.parse_expr()?;
                self.expect(&Token::Comma)?;
                let b = self.parse_expr()?;
                Ok(Expr::Mul(Box::new(a), Box::new(b)))
            }
            Token::Div => {
                self.advance();
                let a = self.parse_expr()?;
                self.expect(&Token::Comma)?;
                let b = self.parse_expr()?;
                Ok(Expr::Div(Box::new(a), Box::new(b)))
            }
            Token::Lt => {
                self.advance();
                let a = self.parse_expr()?;
                self.expect(&Token::Comma)?;
                let b = self.parse_expr()?;
                Ok(Expr::Lt(Box::new(a), Box::new(b)))
            }
            Token::Gt => {
                self.advance();
                let a = self.parse_expr()?;
                self.expect(&Token::Comma)?;
                let b = self.parse_expr()?;
                Ok(Expr::Gt(Box::new(a), Box::new(b)))
            }
            Token::Eq => {
                self.advance();
                let a = self.parse_expr()?;
                self.expect(&Token::Comma)?;
                let b = self.parse_expr()?;
                Ok(Expr::Eq(Box::new(a), Box::new(b)))
            }
            Token::Call => {
                self.advance();
                let name = match self.advance() {
                    Token::Ident(s) => s,
                    Token::Func(s) => s,
                    t => return Err(format!("Expected function name after call, got {t:?}")),
                };
                let args = self.parse_args()?;
                Ok(Expr::Call(name, args))
            }
            Token::Alloc => {
                self.advance();
                let size = self.parse_expr()?;
                Ok(Expr::Alloc(Box::new(size)))
            }
            Token::Int(n) => {
                let n = n;
                self.advance();
                Ok(Expr::Int(n))
            }
            Token::Str(s) => {
                let s = s.clone();
                self.advance();
                Ok(Expr::Str(s))
            }
            Token::Reg(r) => {
                let r = r.clone();
                self.advance();
                if self.peek() == &Token::Dot {
                    self.advance();
                    let field = self.expect_ident()?;
                    return Ok(Expr::Field(r, field));
                }
                Ok(Expr::Reg(r))
            }
            Token::Ident(name) => {
                let name = name.clone();
                self.advance();
                if self.peek() == &Token::LBrace {
                    self.advance();
                    let mut fields = Vec::new();
                    while self.peek() != &Token::RBrace {
                        let fname = self.expect_ident()?;
                        self.expect(&Token::Colon)?;
                        let val = self.parse_expr()?;
                        fields.push((fname, val));
                    }
                    self.expect(&Token::RBrace)?;
                    return Ok(Expr::StructLit(name, fields));
                }
                Ok(Expr::Named(name))
            }
            t => Err(format!("Unexpected token in expression: {t:?}")),
        }
    }

    fn parse_stmt(&mut self) -> Result<Stmt, String> {
        match self.peek().clone() {
            Token::Reg(r) => {
                let r = r.clone();
                self.advance();
                self.expect(&Token::Assign)?;
                let expr = self.parse_expr()?;
                Ok(Stmt::Assign(r, expr))
            }
            Token::Set => {
                self.advance();
                let r = self.expect_reg()?;
                self.expect(&Token::Dot)?;
                let field = self.expect_ident()?;
                self.expect(&Token::Assign)?;
                let expr = self.parse_expr()?;
                Ok(Stmt::SetField(r, field, expr))
            }
            Token::Call => {
                self.advance();
                let name = match self.advance() {
                    Token::Ident(s) => s,
                    Token::Func(s) => s,
                    t => return Err(format!("Expected function name after call, got {t:?}")),
                };
                let args = self.parse_args()?;
                Ok(Stmt::Call(name, args))
            }
            Token::Ret => {
                self.advance();
                let expr = self.parse_expr()?;
                Ok(Stmt::Ret(expr))
            }
            Token::While => {
                self.advance();
                self.expect(&Token::LParen)?;
                let cond = self.parse_expr()?;
                self.expect(&Token::RParen)?;
                self.expect(&Token::LBrace)?;
                let mut body = Vec::new();
                while self.peek() != &Token::RBrace {
                    body.push(self.parse_stmt()?);
                }
                self.expect(&Token::RBrace)?;
                Ok(Stmt::While(cond, body))
            }
            t => Err(format!("Unexpected token in statement: {t:?}")),
        }
    }

    fn parse_struct(&mut self) -> Result<StructDef, String> {
        self.expect(&Token::Struct)?;
        let name = self.expect_ident()?;
        self.expect(&Token::LBrace)?;
        let mut fields = Vec::new();
        while self.peek() != &Token::RBrace {
            let fname = self.expect_ident()?;
            self.expect(&Token::Colon)?;
            let ty = self.parse_ty()?;
            fields.push((fname, ty));
        }
        self.expect(&Token::RBrace)?;
        Ok(StructDef { name, fields })
    }

    fn parse_function(&mut self) -> Result<Function, String> {
        let name = self.expect_func()?;
        let mut params = Vec::new();

        if self.peek() == &Token::LParen {
            self.advance();
            while self.peek() != &Token::RParen {
                let pname = self.expect_reg()?;
                self.expect(&Token::Colon)?;
                let ty = self.parse_ty()?;
                params.push((pname, ty));
                if self.peek() == &Token::Comma {
                    self.advance();
                }
            }
            self.expect(&Token::RParen)?;
        }

        let ret_ty = if self.peek() == &Token::Colon {
            self.advance();
            self.parse_ty()?
        } else {
            Ty::Void
        };

        self.expect(&Token::LBrace)?;
        let mut body = Vec::new();
        while self.peek() != &Token::RBrace {
            body.push(self.parse_stmt()?);
        }
        self.expect(&Token::RBrace)?;

        Ok(Function {
            name,
            params,
            ret_ty,
            body,
        })
    }

    fn parse_program(&mut self) -> Result<Program, String> {
        let mut structs = Vec::new();
        let mut functions = Vec::new();

        while self.peek() != &Token::Eof {
            match self.peek() {
                Token::Struct => structs.push(self.parse_struct()?),
                Token::Func(_) => functions.push(self.parse_function()?),
                t => return Err(format!("Unexpected top-level token: {t:?}")),
            }
        }

        Ok(Program { structs, functions })
    }
}

pub fn parse(tokens: Vec<Token>) -> Result<Program, String> {
    let mut p = Parser { tokens, pos: 0 };
    p.parse_program()
}

#[cfg(test)]
mod tests {
    use super::*;
    use crate::lexer::tokenize;

    fn parse_src(src: &str) -> Program {
        let tokens = tokenize(src).expect("lex failed");
        parse(tokens).expect("parse failed")
    }

    #[test]
    fn minimal_void_function_no_parens() {
        let p = parse_src("@other_func: void { ret 0 }");
        assert_eq!(p.functions.len(), 1);
        assert_eq!(p.functions[0].name, "other_func");
        assert_eq!(p.functions[0].ret_ty, Ty::Void);
        assert!(p.functions[0].params.is_empty());
    }

    #[test]
    fn function_with_empty_parens_and_return_type() {
        let p = parse_src("@main(): int { ret 0 }");
        assert_eq!(p.functions[0].ret_ty, Ty::Int);
    }

    #[test]
    fn call_without_parens() {
        let p = parse_src("@main(): int { call other_func ret 0 }");
        let body = &p.functions[0].body;
        assert!(
            matches!(&body[0], Stmt::Call(name, args) if name == "other_func" && args.is_empty())
        );
    }

    #[test]
    fn call_with_args() {
        let p = parse_src(r#"@main(): int { call puts("hi") ret 0 }"#);
        let body = &p.functions[0].body;
        assert!(matches!(&body[0], Stmt::Call(name, args) if name == "puts" && args.len() == 1));
    }

    #[test]
    fn assign_const_int() {
        let p = parse_src("@main(): int { %x = const 42 ret 0 }");
        assert!(matches!(&p.functions[0].body[0], Stmt::Assign(r, _) if r == "x"));
    }

    #[test]
    fn assign_add() {
        let p = parse_src("@main(): int { %r = add %a, %b ret 0 }");
        assert!(matches!(&p.functions[0].body[0], Stmt::Assign(r, Expr::Add(..)) if r == "r"));
    }

    #[test]
    fn while_loop() {
        let p = parse_src("@main(): int { while (lt %i, 10) { call puts(\"x\") } ret 0 }");
        assert!(matches!(&p.functions[0].body[0], Stmt::While(..)));
    }

    #[test]
    fn set_field() {
        let p = parse_src("@main(): int { set %p.age = const 1 ret 0 }");
        assert!(
            matches!(&p.functions[0].body[0], Stmt::SetField(r, f, _) if r == "p" && f == "age")
        );
    }

    #[test]
    fn field_access_expr() {
        let p = parse_src("@main(): int { %v = %p.name ret 0 }");
        assert!(
            matches!(&p.functions[0].body[0], Stmt::Assign(_, Expr::Field(r, f)) if r == "p" && f == "name")
        );
    }

    #[test]
    fn alloc_expr() {
        let p = parse_src("@main(): int { %buf = alloc 1024 ret 0 }");
        assert!(matches!(
            &p.functions[0].body[0],
            Stmt::Assign(_, Expr::Alloc(_))
        ));
    }

    #[test]
    fn named_constant_read() {
        let p = parse_src("@main(): int { %m = READ ret 0 }");
        assert!(matches!(&p.functions[0].body[0], Stmt::Assign(_, Expr::Named(n)) if n == "READ"));
    }

    #[test]
    fn struct_definition() {
        let p = parse_src("struct Person { name: string age: int }");
        assert_eq!(p.structs.len(), 1);
        assert_eq!(p.structs[0].name, "Person");
        assert_eq!(p.structs[0].fields.len(), 2);
        assert_eq!(p.structs[0].fields[0], ("name".into(), Ty::Str));
        assert_eq!(p.structs[0].fields[1], ("age".into(), Ty::Int));
    }

    #[test]
    fn struct_literal_in_assign() {
        let p = parse_src(r#"@main(): int { %p = Person { name: "Alice" age: 25 } ret 0 }"#);
        assert!(matches!(
            &p.functions[0].body[0],
            Stmt::Assign(_, Expr::StructLit(..))
        ));
    }

    #[test]
    fn parses_simple_bear() {
        let src = r#"
@other_func: void {
    call puts("here i go, doing some thing")
}
@main(): int {
    %0 = const 10
    call puts(%0)
    %1 = const "Hello, world."
    call puts(%1)
    call puts("Hi there.")
    call other_func
    ret 0
}
"#;
        let p = parse_src(src);
        assert_eq!(p.functions.len(), 2);
        assert_eq!(p.functions[0].name, "other_func");
        assert_eq!(p.functions[1].name, "main");
        assert_eq!(p.functions[1].body.len(), 7);
    }

    #[test]
    fn parses_loop_bear() {
        let src = r#"
@main(): int {
    %i = const 0
    while (lt %i, 10) {
        call puts("looping")
        %i = add %i, 1
    }
    ret 0
}
"#;
        let p = parse_src(src);
        assert_eq!(p.functions[0].body.len(), 3);
        assert!(matches!(&p.functions[0].body[1], Stmt::While(..)));
    }

    #[test]
    fn parses_structs_bear() {
        let src = r#"
struct Person {
    name: string
    age: int
}
@main(): int {
    %p = Person { name: "Alice" age: 25 }
    call puts(%p.name)
    set %p.age = add %p.age, 1
    call puts(%p.age)
    ret 0
}
"#;
        let p = parse_src(src);
        assert_eq!(p.structs.len(), 1);
        assert_eq!(p.functions.len(), 1);
    }

    #[test]
    fn unknown_top_level_token_is_error() {
        let tokens = tokenize("42").unwrap();
        assert!(parse(tokens).is_err());
    }
}
